### PR TITLE
MC - parallel functions 

### PR DESCRIPTION
**Pull request by Walidmash**
relates #630 
- Adding process duration timer in the code.
- Adding description to the callback expected from the parallel function.
- Adding notes in NOTES.md for the mentors on MC-Parallel functions.
- Edit the solution so ```counter``` counts the number of unfinished tasks.

### CODE REVIEW

File: coursebook/week-3/NOTES.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+**Mentors Notes:**
+
+## Parallel Functions Morning Challenge
+- Mentor should code along with students, so they can understand the concept of parallel functions and handling asynchronous functions.
```
Comment (eliascodes) [inline coursebook/week-3/NOTES.md]: Does this mean that you intend to run the morning challenge as a code-along? So the students won't be trying to solve it themselves (in pairs)?

File: coursebook/week-3/parallel-solution.js [hunk 2]
```diff
@@ -1,8 +1,8 @@
 function parallel(tasks, callback) {
   // create array to hold results
   var resultsArr = [];
-  // counter to count the loop as resultsArr.length will be misleading
-  var counter = 0;
+  // number of the unfinished tasks
```
Comment (eliascodes) [inline coursebook/week-3/parallel-solution.js]: "the number of unfinished tasks"

File: coursebook/week-3/morning-challenge.md [hunk 3]
```diff
@@ -14,6 +14,8 @@ Complete the 'parallel' function. It should fire all the tasks at the same time,
 
 In case of an error, the final callback should be fired immediately.
 
+The callback accepts two arguments which are error and an array of tasks' results.
```
Comment (eliascodes) [inline coursebook/week-3/morning-challenge.md]: "...array of task results"

File: coursebook/week-3/NOTES.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+- The solution should look like in [parallel-solution](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/week-3/parallel-solution.js), but without ```var hasFailed = false```, students should guess the error that will occur when we don't use it,the solution will work without it in case of one failure task(as in the example), so mentor need to add another failure task to show the problem in case we don't use ```hasFailed```.
```
Comment (eliascodes) [inline coursebook/week-3/NOTES.md]: Thanks for spotting this. 

Actually the solution only "works" in this case because of where we put the line `counter--` (or `counter++`). Because it's below the error check, the counter value at the end of the loop will be `NUMBER_OF_TASKS - NUMBER_OF_ERRORS < tasks.length` in the case of an incrementing counter, or `NUMBER_OF_ERRORS > 0` in the case of a decrementing counter. 

This is a case of us being saved by our own mistakes, rather than actually writing code that works the way we intend 😆 

IMO, the role of the counter should be to keep track of how many tasks have _passed control back to the parallel function_ (which is different to keeping track of how many tasks have successfully completed). Then the `hasFailed` flag will be meaningful in the way that we originally intended.

TLDR:
* Move `counter--` to the top of the `task` callback
* We can get rid of this bullet point

File: coursebook/week-3/NOTES.md (hunk 1 above)
Comment (Walidmash) [reply coursebook/week-3/NOTES.md]: I mean solving it after the students try to solve it themselves, not just viewing solutions, but teaching them the way to start a logical solution to a asynchronous problem.

Comment (eliascodes) [reply coursebook/week-3/NOTES.md]: Ah, OK. 

I think it's valuable to first talk through the solutions that the students come up with though. That will give the mentors a much better understanding of the students level of understanding; which parts they understand, which parts they are struggling with. 

After that discussion, you can better decide where to focus your time when explaining/coding the solution.

Comment (Walidmash) [reply coursebook/week-3/NOTES.md]: Yea
